### PR TITLE
Add PWM Input for STM32

### PR DESCRIPTION
**Pull request by brunob45**
Added `PwmInput` as proposed by @romainreignier in #1954.
Added examples for STM32F446 and STM32F103.
Fixed `SimpleCapture` for STM32F103.

### CODE REVIEW

File: examples/stm32f1/src/bin/pwm_input.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+use defmt::*;
+use embassy_executor::Spawner;
+use embassy_stm32::gpio::{Level, Output, Pull, Speed};
+use embassy_stm32::time::khz;
+use embassy_stm32::timer::pwm_input::PwmInput;
+use embassy_stm32::{bind_interrupts, peripherals, timer};
+use embassy_time::Timer;
+use {defmt_rtt as _, panic_probe as _};
+
+/// Connect PB2 and PB10 with a 1k Ohm resistor
```
Comment (romainreignier) [inline examples/stm32f1/src/bin/pwm_input.rs:14]: Should be

```suggestion
/// Connect PA2 and PC13 with a 1k Ohm resistor
```

File: examples/stm32f1/src/bin/pwm_input.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    let mut pwm_input = PwmInput::new(p.TIM2, p.PA0, Pull::None, khz(10));
+    pwm_input.enable();
+
+    loop {
+        Timer::after_millis(500).await;
+        let _per = pwm_input.get_period_ticks();
+        let _dc = pwm_input.get_duty_ticks();
+        let _pc = pwm_input.get_duty_cycle();
+        asm::nop();
```
Comment (romainreignier) [inline examples/stm32f1/src/bin/pwm_input.rs:50]: Newbie question, why is this line required?

File: examples/stm32f1/src/bin/pwm_input.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    info!("Hello World!");
+
+    unwrap!(spawner.spawn(blinky(p.PC13)));
+
+    let mut pwm_input = PwmInput::new(p.TIM2, p.PA0, Pull::None, khz(10));
+    pwm_input.enable();
+
+    loop {
+        Timer::after_millis(500).await;
+        let _per = pwm_input.get_period_ticks();
```
Comment (romainreignier) [inline examples/stm32f1/src/bin/pwm_input.rs:47]: I think we should have async functions to block on *pulse*, fired by interrupts. Like:
```rs
pwm_input.wait_for_pulse().await;
// or
pwm_input.wait_for_capture().await;
// or
pwm_input.wait_for_period().await;
// and
pwm_input.wait_for_pulse_width().await;
```
In order to be able to retrieve the pulse width of exactly the next pulse (my ultrasonic sensor use case)

File: examples/stm32f4/src/bin/pwm_input.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+use defmt::*;
+use embassy_executor::Spawner;
+use embassy_stm32::gpio::{Level, Output, Pull, Speed};
+use embassy_stm32::time::khz;
+use embassy_stm32::timer::pwm_input::PwmInput;
+use embassy_stm32::{bind_interrupts, peripherals, timer};
+use embassy_time::Timer;
+use {defmt_rtt as _, panic_probe as _};
+
+/// Connect PB2 and PB10 with a 1k Ohm resistor
```
Comment (romainreignier) [inline examples/stm32f4/src/bin/pwm_input.rs:14]:
```suggestion
/// Connect PB2 and PA6 with a 1k Ohm resistor
```

File: embassy-stm32/src/timer/pwm_input.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        self.inner.get_channel_enable_state(Channel::Ch1)
+    }
+
+    /// Get the period tick count
+    pub fn get_period_ticks(&self) -> u32 {
+        self.inner.get_capture_value(self.channel)
+    }
+
+    /// Get the duty tick count
+    pub fn get_duty_ticks(&self) -> u32 {
```
Comment (romainreignier) [inline embassy-stm32/src/timer/pwm_input.rs:118]: I think this should be renamed `get_width_ticks` because it is not yet a duty cycle.

File: examples/stm32f1/src/bin/input_capture.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    let p = embassy_stm32::init(Default::default());
+    info!("Hello World!");
+
+    unwrap!(spawner.spawn(blinky(p.PC13)));
+
+    let ch3 = CapturePin::new_ch3(p.PA2, Pull::None);
+    let mut ic = InputCapture::new(p.TIM2, None, None, Some(ch3), None, Irqs, khz(1000), Default::default());
+
+    loop {
+        info!("wait for risign edge");
```
Comment (romainreignier) [inline examples/stm32f1/src/bin/input_capture.rs:46]:
```suggestion
        info!("wait for rising edge");
```
The same for the F4 example

File: examples/stm32f4/src/bin/pwm_input.rs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    unwrap!(spawner.spawn(blinky(p.PB2)));
+
+    let mut pwm_input = PwmInput::new(p.TIM3, p.PA6, Pull::None, khz(10));
+    pwm_input.enable();
+
+    loop {
+        Timer::after_millis(500).await;
+        let _per = pwm_input.get_period_ticks();
+        let _dc = pwm_input.get_duty_ticks();
+        let _pc = pwm_input.get_duty_cycle();
```
Comment (romainreignier) [inline examples/stm32f4/src/bin/pwm_input.rs:49]: Adding a line like this might help:
```rs
info!("period ticks: {} width ticks: {} duty cycle: {}", per, dc, pc);
```

**Review by romainreignier — COMMENTED** [review]
I have tested both `input_capture` and `pwm_input` on both STM32F103 and STM32F411 and all works.

File: embassy-stm32/src/timer/pwm_input.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
+//! Input capture driver.
```
Comment (romainreignier) [inline embassy-stm32/src/timer/pwm_input.rs:1]:
```suggestion
//! PWM Input driver.
```

File: embassy-stm32/src/timer/pwm_input.rs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+use embassy_hal_internal::into_ref;
+
+use super::low_level::{CountingMode, InputCaptureMode, InputTISelection, Timer};
+use super::{Channel, Channel1Pin, Channel2Pin, GeneralInstance4Channel};
+use crate::gpio::{AFType, Pull};
+use crate::time::Hertz;
+use crate::Peripheral;
+
+/// Input capture driver.
```
Comment (romainreignier) [inline embassy-stm32/src/timer/pwm_input.rs:11]:
```suggestion
/// PWM Input driver.
```

File: embassy-stm32/src/timer/pwm_input.rs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    inner: Timer<'d, T>,
+}
+
+/// Convert pointer to TIM instance to TimGp16 object
+fn regs_gp16(ptr: *mut ()) -> crate::pac::timer::TimGp16 {
+    unsafe { crate::pac::timer::TimGp16::from_ptr(ptr) }
+}
+
+impl<'d, T: GeneralInstance4Channel> PwmInput<'d, T> {
+    /// Create a new input capture driver.
```
Comment (romainreignier) [inline embassy-stm32/src/timer/pwm_input.rs:23]:
```suggestion
    /// Create a new PWM input driver.
```

File: embassy-stm32/src/timer/pwm_input.rs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        critical_section::with(|_| {
+            pin.set_as_af_pull(pin.af_num(), AFType::Input, pull_type);
+            #[cfg(gpio_v2)]
+            pin.set_speed(crate::gpio::Speed::VeryHigh);
+        });
+
+        Self::new_inner(tim, freq, Channel::Ch1, Channel::Ch2)
+    }
+
+    /// Create a new input capture driver.
```
Comment (romainreignier) [inline embassy-stm32/src/timer/pwm_input.rs:40]:
```suggestion
    /// Create a new PWM input driver.
```
And it should be documented that in this case, the Channel 2 GPIO is the input of the signal if I understand correctly.

**Review by romainreignier — COMMENTED** [review]
Also tested the `new_alt`on F1 with success.

File: examples/stm32f1/src/bin/pwm_input.rs (hunk 2 above)
Comment (brunob45) [reply examples/stm32f1/src/bin/pwm_input.rs:50]: It was used to put a break point. I will remove it when the PR is ready

**Comment by brunob45** [comment]
@romainreignier thank you for the tests and reviews! I'll keep working on this PR before I mark it as "ready"

File: embassy-stm32/src/timer/input_capture.rs [hunk 12]
```diff
@@ -40,11 +40,9 @@ macro_rules! channel_impl {
...
             pub fn $new_chx(pin: impl Peripheral<P = impl $pin_trait<T>> + 'd, pull_type: Pull) -> Self {
                 into_ref!(pin);
-                critical_section::with(|_| {
-                    pin.set_as_af_pull(pin.af_num(), AFType::Input, pull_type);
-                    #[cfg(gpio_v2)]
-                    pin.set_speed(crate::gpio::Speed::VeryHigh);
-                });
+
+                pin.set_as_af_pull(pin.af_num(), AFType::Input, pull_type);
+
```
Comment (brunob45) [inline embassy-stm32/src/timer/input_capture.rs:45]: remove set_speed as it only affects outputs

File: embassy-stm32/src/timer/input_capture.rs [hunk 13]
```diff
@@ -109,24 +107,6 @@ impl<'d, T: GeneralInstance4Channel> InputCapture<'d, T> {
         self.inner.get_channel_enable_state(channel)
     }
 
-    /// Set tick frequency.
-    ///
-    /// Note: when you call this, the max period value changes
-    pub fn set_tick_freq(&mut self, freq: Hertz) {
```
Comment (brunob45) [inline embassy-stm32/src/timer/input_capture.rs:115]: function moved to `inner`

**Comment by brunob45** [comment]
Async functions will be done in another PR

**Comment by romainreignier** [comment]
The Input Capture fix might go in a separate PR as well to get merges faster.
Anyway, I have managed to get distance readings from my US sensor with the Input Capture module as you told me: 
https://github.com/romainreignier/bluepill-tank-level/blob/main/src/main.rs
Thanks

**Comment by Dirbaio** [comment]
(ci got stuck)

bender run

**Review by Dirbaio — APPROVED** [review]
thanks!
